scripts/2015_nips_paper/plot/plot_ranks.py

```diff
@@ -0,0 +1,176 @@
+    series = series.fillna(method='ffill')
+
+    # return the trajectories over seeds (series object)
+    return series
+
+
+def main():
+    # name of the file where the plot is stored
+    saveto = "../plot.png"
+    # runtime of each experiment
+    max_runtime = 60
+    # folder where all trajectories are stored.
+    working_directory = "../log_output"
+
+    # list of models
+    model_list = ['vanilla', 'ensemble', 'metalearning', 'meta_ensemble']
+
+    # list of seeds
+    seed_dir = os.path.join(working_directory, 'vanilla')
+    seed_list = [seed for seed in os.listdir(seed_dir)]
+
+    # list of tasks
+    vanilla_task_dir = os.path.join(seed_dir, seed_list[0])
+    task_list = [task_id for task_id in os.listdir(vanilla_task_dir)]
+
+    # Step 1. Merge all trajectories into one Dataframe object.
+    #####################################################################################
+    trajectories_all_models = []
+    for model in model_list:
+        trajectories = []
+        for task_id in task_list:
+            csv_files = []
+
+            for seed in seed_list:
+                # collect all csv files of different seeds for current model and
+                # current task.
+                if model in ['vanilla', 'ensemble']:
+                    csv_file = os.path.join(working_directory,
+                                            'vanilla',
+                                            seed,
+                                            task_id,
+                                            "score_{}.csv".format(model)
+                                            )
+
+                elif model in ['metalearning', 'meta_ensemble']:
+                    csv_file = os.path.join(working_directory,
+                                            'metalearning',
+                                            seed,
+                                            task_id,
+                                            "score_{}.csv".format(model),
+                                            )
+                csv_files.append(csv_file)
+
+            performance_list = []
+            time_list = []
+
+            # Get data from csv
+            for fl in csv_files:
+                _, csv_data = read_csv(fl, has_header=True)
+                csv_data = np.array(csv_data)
+                # Replace too high values with args.maxsize
+                data = [min([sys.maxsize, float(i.strip())]) for i in
+                        csv_data[:, 2]]  # test trajectories are stored in third column
+
+                time_steps = [float(i.strip()) for i in csv_data[:, 0]]
+                assert time_steps[0] == 0
+
+                performance_list.append(data)
+                time_list.append(time_steps)
+
+            # trajectory is the pd.Series object containing all seed runs of the
+            # current model and current task.
+            trajectory = fill_trajectory(performance_list, time_list)

```

Comment:
Don't you have to do this once the data for all models is loaded, i.e. outside the nested for loops?

Reply:
Do I understand you correctly that the code should fill all trajectories across all models to share the same time step indices? That happens in line 142-146, where fillna() is applied to the picked trajectories, and in line 156, where fillna() is applied to the computed ranks. This comes mostly from the code you sent me, which I admittedly tried not to alter too much. However, I see that it makes more sense and intuitive to do fill_trajectory once and for all models when creating the dataframe object, instead of filling the trajectories many times later. Should I modify the code in such a way?

Reply:
Okay, thanks for the explanation, that makes sense.